#160618108 Star rating on articles

VictorUkafor opened the pull request:
### What does this PR do?
This pull request allows users to star rate articles

### Description of Task to be completed?
- Write tests for the  star rating features
- Write middleware for the fetching a single article
- Write middleware for checking if the user has rating article before
- Write method for implementing star rating feature

### Any background context you want to provide?
N/A

### What are the relevant pivotal tracker stories?
#160618108

### Screenshots (if appropriate)
N/A

### Questions:
None

AdinoyiSadiq commented on controllers/articleController.js:
```diff
@@ -165,11 +163,45 @@ const reactToArticle = async (req, res) => {
+    error: 'Internal server error'
+  });
+};
+
```
Hi Victor, this looks great. However, I think we should use the 'async await' pattern whenever possible. Also, I would suggest using the 'try and catch' pattern as well.

AdinoyiSadiq commented on controllers/articleController.js:
```diff
@@ -165,11 +163,45 @@ const reactToArticle = async (req, res) => {
+  const postedRating = await helper.addRating(req.userRating, req.userId, req.article.id);
+  return addRatingsToArticle(res, req, postedRating);
+};
+
```
Hi Victor. Firstly I like how succinct this function is.

I have a suggestion though. I think addRatingsToArticle functionality shouldn't have 'res, req' passed into it as arguments. From experience, this might lead to errors with node where it tries to send a response twice. 

I would suggest having the function return a value which is then tested by a conditional statement, before returning a response. 

TLDR All responses should be sent from your controller in my opinion.

AdinoyiSadiq commented on routes/api/articles.js:
```diff
@@ -9,7 +14,7 @@ router.put('/articles/:slug', authenticate, checkNullInput, articleController.up
 router.get('/articles', authenticate, articleController.getAllArticles);
 router.get('/articles/:slug', authenticate, articleController.getAnArticle);
+router.post('/articles/:articleId/rating', authenticate, checkIfArticleExist, getUserRating, articleController.postRating);
 router.post('/articles/:slug/react', authenticate, reactionValidation, articleController.reactToArticle);
```
So, I would suggest that we use only one validator here. Maybe combine checkIfArticleExists and get getUserRating. If getUserRating is not validating the data that has been posted then I think that it should be in your controller instead.

 TLDR I would have one validator here and move the getUserRating function call to the 'postRating' controller.

AdinoyiSadiq commented on routes/api/articles.js:
```diff
@@ -9,7 +14,7 @@ router.put('/articles/:slug', authenticate, checkNullInput, articleController.up
 router.delete('/articles/:slug', authenticate, articleController.deleteArticle);
 router.get('/articles', authenticate, articleController.getAllArticles);
 router.get('/articles/:slug', authenticate, articleController.getAnArticle);
+router.post('/articles/:articleId/rating', authenticate, checkIfArticleExist, getUserRating, articleController.postRating);
```
So, I would suggest that we use only one validator here. Maybe combine checkIfArticleExists and get getUserRating. If getUserRating is not validating the data that has been posted then I think that it should be in your controller instead.

 TLDR I would have one validator here and move the getUserRating function call to the 'postRating' controller.

VictorUkafor replied:
Thank you for pointing this point out I've fixed it already

Commit:
feat(star-rating): add logic for star rating article

- add middleware for validating inputs
- add controller logic for implementing star rating feature

[Delivers #160618105]

Commit:
feat(star-rating): add more test caes

- add more test cases to increase test coverage

[Delivers #160618105]

Commit:
feat(star-rating): add more test caes

- add more test cases to increase test coverage

[Delivers #160618105]

Commit:
feat(star-rating): add more test caes

- add more test cases to increase test coverage

[Delivers #160618105]

Commit:
feat(star-rating): fix merge conflict

- fix merge conflict

[Delivers #160618105]

Commit:
feat(star-rating): fix merge conflict

- fix merge conflict

[Delivers #160618105]

Commit:
feat(star-rating): fix merge conflict

- fix merge conflict

[Delivers #160618105]

Commit:
feat(star-rating): fix merge conflict

- fix merge conflict

[Delivers #160618105]
